sharedbuffers/pool.py

```diff
@@ -0,0 +1,175 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import tempfile
+import mmap
+import struct
+
+from .mapped_struct import StrongIdMap
+
+# Default section size is set to 128MB which is a size at which most
+# malloc implementations turn to mmap
+DEFAULT_SECTION_SIZE = 128<<20
+
+DEFAULT_PACK_BUFFER = 65536
+
+class Section(object):
+
+    def __init__(self, buf, implicit_offs=0, idmap_kwargs={}):
+        self.buf = buf
+        self.real_buf = buffer(buf)
+        self.implicit_offs = implicit_offs
+        self.write_pos = 0
+        self.idmap = StrongIdMap(**idmap_kwargs)
+
+    def allocate(self, size=None):
+        write_pos = self.write_pos
+        if size is None:
+            size = len(self.buf) - write_pos
+        if (write_pos + size) <= len(self.buf):
+            self.write_pos += size
+            return write_pos
+        else:
+            raise IndexError("Buffer overflow trying to allocate %d bytes from section" % size)
+
+    def resize(self, pos, new_size):
+        new_end = pos + new_size
+        if new_end != self.write_pos:
+            raise RuntimeError("Cannot resize non-tail buffers")
+        self.write_pos = new_end
+
+    def append(self, buf, verify_pos=None):
+        write_pos = self.allocate(len(buf))
+        if verify_pos is not None and verify_pos != write_pos:
+            raise RuntimeError("Concurrent modification")
+        self.buf[write_pos:write_pos+len(buf)] = bytes(buf)
+        return write_pos
+
+    @property
+    def free_space(self):
+        return len(self.buf) - self.write_pos
+
+class BaseObjectPool(object):
+
+    def _mktemp(self):
+        raise NotImplementedError
+
+    def __init__(self, section_size=DEFAULT_SECTION_SIZE, temp_kwargs={}, idmap_kwargs={}):
+        self.temp_kwargs = temp_kwargs
+        self.idmap_kwargs = idmap_kwargs
+        self.section_size = section_size
+        self.sections = []
+        self.freehead = 0
+        self.total_size = 0
+        self.idmap_preload = []
+
+    @property
+    def size(self):
+        if self.sections:
+            last_section = self.sections[-1]
+            return last_section.implicit_offs + last_section.write_pos
+        else:
+            return 0
+
+    def add_section(self):
+        f = self._mktemp()
+        try:
+            f.truncate(self.section_size)
+            buf = mmap.mmap(
+                f.fileno(), 0,
+                access = mmap.ACCESS_WRITE)
```

Comment:
Is this buffer going to be shared among processes ? If so, you need to call ```fsync``` periodically to make sure memory state is correctly synchronized.

Reply:
Memory is always synchronized. `msync` (not fsync) would be needed to make sure the changes hit the disk. We don't currently require this.